src/tbb/threading_control.cpp

```diff
@@ -0,0 +1,397 @@
+    my_cancellation_disseminator->unregister_thread(td);
+}
+
+void threading_control_impl::propagate_task_group_state(std::atomic<uint32_t> d1::task_group_context::*mptr_state,
+                                d1::task_group_context& src, uint32_t new_state)
+{
+    my_cancellation_disseminator->propagate_task_group_state(mptr_state, src, new_state);
+}
+
+std::size_t threading_control_impl::worker_stack_size() {
+    return my_thread_dispatcher->worker_stack_size();
+}
+
+unsigned threading_control_impl::max_num_workers() {
+    return my_thread_dispatcher->my_num_workers_hard_limit;
+}
+
+bool threading_control_impl::check_client_priority(threading_control_impl::threading_control_client client) {
+    return static_cast<pm_client*>(client.first)->is_top_priority();
+}
+
+void threading_control_impl::adjust_demand(threading_control_impl::threading_control_client tc_client, int mandatory_delta, int workers_delta) {
+    pm_client& c = *static_cast<pm_client*>(tc_client.first);
+    my_thread_request_serializer->register_mandatory_request(mandatory_delta);
+    my_permit_manager->adjust_demand(c, mandatory_delta, workers_delta);
+}
+
+thread_control_monitor& threading_control_impl::get_waiting_threads_monitor() {
+    return *my_waiting_threads_monitor;
+}
+
+// ---------------------------------------- threading_control -------------------------------------------------------------------
+
+void global_control_lock();
+void global_control_unlock();
+
+void threading_control::add_ref(bool is_public) {
+    ++my_ref_count;
+    if (is_public) {
+        my_public_ref_count++;
+    }
+}
+
+bool threading_control::remove_ref(bool is_public) {
+    if (is_public) {
+        __TBB_ASSERT(g_threading_control == this, "Global threading control instance was destroyed prematurely?");
+        __TBB_ASSERT(my_public_ref_count.load(std::memory_order_relaxed), nullptr);
+        --my_public_ref_count;
+    }
+
+    bool is_last_ref = --my_ref_count == 0;
+    if (is_last_ref) {
+        __TBB_ASSERT(!my_public_ref_count.load(std::memory_order_relaxed), nullptr);
+        g_threading_control = nullptr;
+    }
+
+    return is_last_ref;
+}
+
+threading_control* threading_control::get_threading_control(bool is_public) {
+    threading_control* control = g_threading_control;
+    if (control) {
+        control->add_ref(is_public);
+    }
+
+    return control;
+}
+
+threading_control* threading_control::create_threading_control() {
+    threading_control* thr_control{ nullptr };
+    try_call([&] {
+        // Global control should be locked before threading_control_impl
+        global_control_lock();
+        global_mutex_type::scoped_lock lock(g_threading_control_mutex);
+
+        thr_control = get_threading_control(/*public = */ true);
+        if (thr_control != nullptr) {
+            return;
+        }
+
+        // POSSIBLE MEMORY LEAK THERE!!!!!!!
+        thr_control = new (cache_aligned_allocate(sizeof(threading_control))) threading_control(1, 1);
```

Comment:
What is `threading_control(1, 1)`?

Reply:
one public and one private reference the go in pair on creation. I just want to expose this logic from default ctor and make it more obvious.